[WebGPU] Make it explicit that __wgslMin is only called on uint parameters

mwyrzykowski opened the pull request:
#### d9758a8fd190860ac6c4a468dfdea906786530e5
<pre>
[WebGPU] Make it explicit that __wgslMin is only called on uint parameters
<a href="https://bugs.webkit.org/show_bug.cgi?id=282937">https://bugs.webkit.org/show_bug.cgi?id=282937</a>
<a href="https://rdar.apple.com/138789162">rdar://138789162</a>

Reviewed by Tadeu Zagallo.

__wgslMin was only ever called with uint arguments so no need
for it to be templated.

Furthermore, we can remove the use of volatile indices by using
select() instead of min().

* LayoutTests/fast/webgpu/nocrash/fuzz-282937-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-282937.html: Added.
Add regression test.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::imageBytesForSource):
(WebCore::GPUQueue::copyExternalImageToTexture):
Test case found web content crash as well, fixing so the test passes.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
Use select() instead of min() to avoid volatile variables.

Canonical link: <a href="https://commits.webkit.org/286488@main">https://commits.webkit.org/286488@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c315b19b0151715b673b4012223746de24a655df

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76013 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55042 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28912 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80510 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27279 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78129 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64184 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3338 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59599 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17755 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79080 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49490 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65284 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39957 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46889 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22768 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25606 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/68005 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23105 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81974 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3382 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2165 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67830 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65252 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67139 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16755 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11092 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9215 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3332 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6138 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3353 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4291 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3360 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

mwyrzykowski commented on Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
```diff
@@ -515,14 +515,11 @@ void FunctionDefinitionWriter::emitNecessaryHelpers()
     }
 
     if (m_shaderModule.usesMin()) {
-        m_stringBuilder.append(m_indent, "template<typename T>\n"_s,
```
we can leave this templated if preferred, but I don't see it used outside of `uint`, can you confirm @tadeuzagallo ?

tadeuzagallo replied:
That looks correct, looks like it's only used in one place

mwyrzykowski committed:
[WebGPU] Make it explicit that __wgslMin is only called on uint parameters
https://bugs.webkit.org/show_bug.cgi?id=282937
rdar://138789162

Reviewed by Tadeu Zagallo.

__wgslMin was only ever called with uint arguments so no need
for it to be templated.

Furthermore, we can remove the use of volatile indices by using
select() instead of min().

* LayoutTests/fast/webgpu/nocrash/fuzz-282937-expected.txt: Added.
* LayoutTests/fast/webgpu/nocrash/fuzz-282937.html: Added.
Add regression test.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::imageBytesForSource):
(WebCore::GPUQueue::copyExternalImageToTexture):
Test case found web content crash as well, fixing so the test passes.

* Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp:
(WGSL::Metal::FunctionDefinitionWriter::emitNecessaryHelpers):
Use select() instead of min() to avoid volatile variables.

Canonical link: https://commits.webkit.org/286488@main
